pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -239,84 +206,14 @@ func (r *ReconcileHyperConverged) Reconcile(request reconcile.Request) (reconcil
-			for i, f := range []func(c client.Client, cr *hcov1alpha1.HyperConverged) error{
-				// Keep ensuring that we are trying to delete
-				// protected resources first
-				ensureKubeVirtDeleted,
-				ensureCDIDeleted,
-				ensureNetworkAddonsDeleted,
-				ensureKubeVirtCommonTemplateBundleDeleted,
-			} {
-				err = f(r.client, instance)
-				if err != nil {
-					reqLogger.Error(err, "Failed to manually delete objects")
-
-					// TODO: ask to other components to expose something like
-					// func IsDeleteRefused(err error) bool
-					// to be able to clearly distinguish between an explicit
-					// refuse from other operator and any other kind of error that
-					// could potentially happen in the process
-					errT := ErrHCOUninstall
-					errMsg := uninstallHCOErrorMsg
-					switch i {
-					case 0:
-						errT = ErrVirtUninstall
-						errMsg = uninstallVirtErrorMsg + err.Error()
-					case 1:
-						errT = ErrCDIUninstall
-						errMsg = uninstallCDIErrorMsg + err.Error()
-					}
-
-					errE := r.emitEvent(instance, reqLogger, corev1.EventTypeWarning, errT, errMsg)
-					if errE != nil {
-						reqLogger.Error(errE, "Failed emitting uninstall error event")
-					}
-
-					// TODO: implement a validating webhook to try to delete virt and CDI CRs
-					// in dry run mode before really accepting the deletion request.
-					// This event should still stay here because no strategy can ensure we are
-					// 100% race conditions free
-
-					return reconcile.Result{}, err
-				}
-			}
-
-			// Remove the finalizer
-			instance.ObjectMeta.Finalizers = drop(instance.ObjectMeta.Finalizers, FinalizerName)
-			err = r.client.Update(context.TODO(), instance)
-
-			// Need to requeue because finalizer update does not change metadata.generation
-			return reconcile.Result{Requeue: true}, err
+			requeue, err := r.recursiveDelete(instance, reqLogger)
+			return reconcile.Result{Requeue: requeue}, err
 		}
 	}
 
-	for _, f := range []func(*hcov1alpha1.HyperConverged, logr.Logger, reconcile.Request) error{
-		r.ensureKubeVirtConfig,
-		r.ensureKubeVirtStorageConfig,
-		r.ensureKubeVirt,
-		r.ensureCDI,
-		r.ensureNetworkAddons,
-		r.ensureKubeVirtCommonTemplateBundle,
-		r.ensureKubeVirtNodeLabellerBundle,
-		r.ensureKubeVirtTemplateValidator,
-		r.ensureKubeVirtMetricsAggregation,
-		r.ensureIMSConfig,
-	} {
-		err = f(instance, reqLogger, request)
-		if err != nil {
-			conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
-				Type:    hcov1alpha1.ConditionReconcileComplete,
-				Status:  corev1.ConditionFalse,
-				Reason:  reconcileFailed,
-				Message: fmt.Sprintf("Error while reconciling: %v", err),
-			})
-			// don't want to overwrite the actual reconcile failure
-			uErr := r.client.Status().Update(context.TODO(), instance)
-			if uErr != nil {
-				reqLogger.Error(uErr, "Failed to update conditions")
-			}
-			return reconcile.Result{}, err
-		}
+	err = r.ensureComponents(instance, request, reqLogger)
```

Comment:
If you want to refactor the `Reconcile()` func, you should do that in a separate PR.